// keras-js/lib/layers/convolutional/SeparableConvolution2D.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _activations = require('../../activations');

var activations = _interopRequireWildcard(_activations);

var _Tensor = require('../../Tensor');

var _Tensor2 = _interopRequireDefault(_Tensor);

var _Layer2 = require('../../Layer');

var _Layer3 = _interopRequireDefault(_Layer2);

var _Convolution2D2 = require('./Convolution2D');

var _Convolution2D3 = _interopRequireDefault(_Convolution2D2);

var _ndarrayOps = require('ndarray-ops');

var _ndarrayOps2 = _interopRequireDefault(_ndarrayOps);

var _ndarrayGemm = require('ndarray-gemm');

var _ndarrayGemm2 = _interopRequireDefault(_ndarrayGemm);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * _DepthwiseConvolution2D layer class
 */
var _DepthwiseConvolution2D = function (_Convolution2D) {
  _inherits(_DepthwiseConvolution2D, _Convolution2D);

  function _DepthwiseConvolution2D() {
    var attrs = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    _classCallCheck(this, _DepthwiseConvolution2D);

    return _possibleConstructorReturn(this, (_DepthwiseConvolution2D.__proto__ || Object.getPrototypeOf(_DepthwiseConvolution2D)).call(this, attrs));
  }

  /**
   * Convert input image to column matrix
   * @param {Tensor} x
   * @returns {Tensor} x
   */


  _createClass(_DepthwiseConvolution2D, [{
    key: '_im2col',
    value: function _im2col(x) {
      var _x$tensor$shape = _slicedToArray(x.tensor.shape, 3),
          inputRows = _x$tensor$shape[0],
          inputCols = _x$tensor$shape[1],
          inputChannels = _x$tensor$shape[2];

      var nbRow = this.kernelShape[1];
      var nbCol = this.kernelShape[2];
      var outputRows = this.outputShape[0];
      var outputCols = this.outputShape[1];
      var nbPatches = outputRows * outputCols;
      var patchLen = nbRow * nbCol;

      if (!this._imColsMat) {
        this._imColsMat = new _Tensor2.default([], [nbPatches * inputChannels, patchLen]);
      }

      var patch = new _Tensor2.default([], [nbRow, nbCol, 1]);
      var offset = 0;
      for (var c = 0; c < inputChannels; c++) {
        for (var i = 0, limit = inputRows - nbRow; i <= limit; i += this.subsample[0]) {
          for (var j = 0, _limit = inputCols - nbCol; j <= _limit; j += this.subsample[1]) {
            _ndarrayOps2.default.assign(patch.tensor, x.tensor.hi(i + nbRow, j + nbCol, c + 1).lo(i, j, c));
            this._imColsMat.tensor.data.set(patch.tensor.data, offset);
            offset += patchLen;
          }
        }
      }

      if (this._useWeblas) {
        this._imColsMat.createWeblasTensor();
      }
      return this._imColsMat;
    }

    /**
     * Convert filter weights to row matrix
     * @returns {Tensor|weblas.pipeline.Tensor} wRowsMat
     */

  }, {
    key: '_w2row',
    value: function _w2row() {
      var inputChannels = this.weights.W.tensor.shape[2];

      var _kernelShape = _slicedToArray(this.kernelShape, 3),
          nbFilter = _kernelShape[0],
          nbRow = _kernelShape[1],
          nbCol = _kernelShape[2];

      var patchLen = nbRow * nbCol;

      this._wRowsMat = new _Tensor2.default([], [patchLen, nbFilter * inputChannels]);

      var patch = new _Tensor2.default([], [nbRow, nbCol]);
      var patchRaveled = new _Tensor2.default([], [patchLen]);
      var p = 0;
      for (var c = 0; c < inputChannels; c++) {
        for (var n = 0; n < nbFilter; n++) {
          _ndarrayOps2.default.assign(patch.tensor, this.weights.W.tensor.pick(null, null, c, n));
          patchRaveled.replaceTensorData(patch.tensor.data);
          _ndarrayOps2.default.assign(this._wRowsMat.tensor.pick(null, p), patchRaveled.tensor);
          p += 1;
        }
      }

      return this._wRowsMat;
    }

    /**
     * Method for layer computational logic
     * @param {Tensor} x
     * @returns {Tensor} x
     */

  }, {
    key: 'call',
    value: function call(x) {
      this._calcOutputShape(x.tensor.shape);
      this._padInput(x);

      this._im2col(x);

      var nbFilter = this.kernelShape[0];
      var outputRows = this.outputShape[0];
      var outputCols = this.outputShape[1];
      var nbPatches = outputRows * outputCols;
      var matMul = new _Tensor2.default([], [nbPatches * x.tensor.shape[2], nbFilter * x.tensor.shape[2]]);

      if (this._useWeblas && !(this._imColsMat._gpuMaxSizeExceeded || this._wRowsMat._gpuMaxSizeExceeded)) {
        // GPU
        matMul.tensor.data = weblas.pipeline.sgemm(1, this._imColsMat.weblasTensor, this._wRowsMat.weblasTensor, 1, this._zerosVec.weblasTensor).transfer();
      } else {
        // CPU
        (0, _ndarrayGemm2.default)(matMul.tensor, this._imColsMat.tensor, this._wRowsMat.tensor, 1, 1);
      }

      var output = new _Tensor2.default([], [outputRows, outputCols, x.tensor.shape[2] * nbFilter]);
      var outputDataLength = outputRows * outputCols * x.tensor.shape[2] * nbFilter;
      var dataFiltered = new Float32Array(outputDataLength);
      for (var c = 0; c < x.tensor.shape[2]; c++) {
        for (var i = 0, n = c * outputDataLength + c * nbFilter, len = (c + 1) * outputDataLength; n < len; i++, n += nbFilter * x.tensor.shape[2]) {
          for (var m = 0; m < nbFilter; m++) {
            dataFiltered[n + m - c * outputDataLength] = matMul.tensor.data[n + m];
          }
        }
      }
      output.replaceTensorData(dataFiltered);

      x.tensor = output.tensor;

      return x;
    }
  }]);

  return _DepthwiseConvolution2D;
}(_Convolution2D3.default);

/**
 * SeparableConvolution2D layer class
 */


var SeparableConvolution2D = function (_Layer) {
  _inherits(SeparableConvolution2D, _Layer);

  /**
   * Creates a SeparableConvolution2D layer
   * @param {number} attrs.nbFilter - Number of convolution filters to use.
   * @param {number} attrs.nbRow - Number of rows in the convolution kernel.
   * @param {number} attrs.nbCol - Number of columns in the convolution kernel.
   * @param {Object} [attrs] - layer attributes
   */
  function SeparableConvolution2D() {
    var attrs = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    _classCallCheck(this, SeparableConvolution2D);

    var _this2 = _possibleConstructorReturn(this, (SeparableConvolution2D.__proto__ || Object.getPrototypeOf(SeparableConvolution2D)).call(this, attrs));

    _this2.layerClass = 'SeparableConvolution2D';

    var _attrs$nbFilter = attrs.nbFilter,
        nbFilter = _attrs$nbFilter === undefined ? 1 : _attrs$nbFilter,
        _attrs$nbRow = attrs.nbRow,
        nbRow = _attrs$nbRow === undefined ? 1 : _attrs$nbRow,
        _attrs$nbCol = attrs.nbCol,
        nbCol = _attrs$nbCol === undefined ? 1 : _attrs$nbCol,
        _attrs$activation = attrs.activation,
        activation = _attrs$activation === undefined ? 'linear' : _attrs$activation,
        _attrs$borderMode = attrs.borderMode,
        borderMode = _attrs$borderMode === undefined ? 'valid' : _attrs$borderMode,
        _attrs$subsample = attrs.subsample,
        subsample = _attrs$subsample === undefined ? [1, 1] : _attrs$subsample,
        _attrs$depthMultiplie = attrs.depthMultiplier,
        depthMultiplier = _attrs$depthMultiplie === undefined ? 1 : _attrs$depthMultiplie,
        _attrs$dimOrdering = attrs.dimOrdering,
        dimOrdering = _attrs$dimOrdering === undefined ? 'tf' : _attrs$dimOrdering,
        _attrs$bias = attrs.bias,
        bias = _attrs$bias === undefined ? true : _attrs$bias;


    _this2.activation = activation;
    _this2.activationFunc = activations[activation];

    if (borderMode === 'valid' || borderMode === 'same') {
      _this2.borderMode = borderMode;
    } else {
      throw new Error(`${_this2.name} [SeparableConvolution2D layer] Invalid borderMode.`);
    }

    _this2.subsample = subsample;
    _this2.depthMultiplier = depthMultiplier;

    if (dimOrdering === 'tf' || dimOrdering === 'th') {
      _this2.dimOrdering = dimOrdering;
    } else {
      throw new Error(`${_this2.name} [SeparableConvolution2D layer] Only tf and th dim ordering are allowed.`);
    }

    _this2.bias = bias;

    // Layer weights specification
    _this2.params = _this2.bias ? ['depthwise_kernel', 'pointwise_kernel', 'b'] : ['depthwise_kernel', 'pointwise_kernel'];

    // SeparableConvolution2D has two components: depthwise, and pointwise.
    // Activation function and bias is applied at the end.
    // Subsampling (striding) only performed on depthwise part, not the pointwise part.
    _this2.depthwiseConvAttrs = {
      nbFilter: _this2.depthMultiplier,
      nbRow,
      nbCol,
      activation: 'linear',
      borderMode,
      subsample,
      dimOrdering,
      bias: false,
      gpu: attrs.gpu
    };
    _this2.pointwiseConvAttrs = {
      nbFilter,
      nbRow: 1,
      nbCol: 1,
      activation: 'linear',
      borderMode,
      subsample: [1, 1],
      dimOrdering,
      bias: _this2.bias,
      gpu: attrs.gpu
    };
    return _this2;
  }

  /**
   * Method for setting layer weights
   * Override `super` method since weights must be set in component Convolution2D layers
   * @param {Tensor[]} weightsArr - array of weights which are instances of Tensor
   */


  _createClass(SeparableConvolution2D, [{
    key: 'setWeights',
    value: function setWeights(weightsArr) {
      this._depthwiseConv = new _DepthwiseConvolution2D(this.depthwiseConvAttrs);
      this._depthwiseConv.setWeights(weightsArr.slice(0, 1));
      this._pointwiseConv = new _Convolution2D3.default(this.pointwiseConvAttrs);
      this._pointwiseConv.setWeights(weightsArr.slice(1, 3));
    }

    /**
     * Method for layer computational logic
     * @param {Tensor} x
     * @returns {Tensor} x
     */

  }, {
    key: 'call',
    value: function call(x) {
      // Perform depthwise ops
      var depthwiseOutput = this._depthwiseConv.call(x);

      // Perform depthwise ops
      var pointwiseOutput = this._pointwiseConv.call(depthwiseOutput);

      x.tensor = pointwiseOutput.tensor;

      // activation
      this.activationFunc(x);

      return x;
    }
  }]);

  return SeparableConvolution2D;
}(_Layer3.default);

exports.default = SeparableConvolution2D;
