// keras-js/lib/layers/recurrent/GRU.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _activations = require('../../activations');

var activations = _interopRequireWildcard(_activations);

var _Tensor = require('../../Tensor');

var _Tensor2 = _interopRequireDefault(_Tensor);

var _Layer2 = require('../../Layer');

var _Layer3 = _interopRequireDefault(_Layer2);

var _ndarrayBlasLevel = require('ndarray-blas-level2');

var _ndarrayOps = require('ndarray-ops');

var _ndarrayOps2 = _interopRequireDefault(_ndarrayOps);

var _cwise = require('cwise');

var _cwise2 = _interopRequireDefault(_cwise);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * GRU layer class
 */
var GRU = function (_Layer) {
  _inherits(GRU, _Layer);

  /**
   * Creates a GRU layer
   * @param {number} attrs.outputDim - output dimensionality
   * @param {number} [attrs.activation] - activation function
   * @param {number} [attrs.innerActivation] - inner activation function
   * @param {number} [attrs.returnSequences] - return the last output in the output sequence or the full sequence
   * @param {number} [attrs.goBackwards] - process the input sequence backwards
   * @param {number} [attrs.stateful] - whether to save the last state as the initial state for the next pass
   * @param {Object} [attrs] - layer attributes
   */
  function GRU() {
    var attrs = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

    _classCallCheck(this, GRU);

    var _this = _possibleConstructorReturn(this, (GRU.__proto__ || Object.getPrototypeOf(GRU)).call(this, attrs));

    _this._combine = (0, _cwise2.default)({
      args: ['array', 'array', 'array', 'array'],
      body: function body(_y, _x1, _x2, _b) {
        _y = _x1 + _x2 + _b;
      }
    });
    _this._update = (0, _cwise2.default)({
      args: ['array', 'array', 'array'],
      body: function body(_h, _htm1, _z) {
        _h = _h * (1 - _z) + _htm1 * _z;
      }
    });

    _this.layerClass = 'GRU';

    var _attrs$outputDim = attrs.outputDim,
        outputDim = _attrs$outputDim === undefined ? 1 : _attrs$outputDim,
        _attrs$activation = attrs.activation,
        activation = _attrs$activation === undefined ? 'tanh' : _attrs$activation,
        _attrs$innerActivatio = attrs.innerActivation,
        innerActivation = _attrs$innerActivatio === undefined ? 'hardSigmoid' : _attrs$innerActivatio,
        _attrs$returnSequence = attrs.returnSequences,
        returnSequences = _attrs$returnSequence === undefined ? false : _attrs$returnSequence,
        _attrs$goBackwards = attrs.goBackwards,
        goBackwards = _attrs$goBackwards === undefined ? false : _attrs$goBackwards,
        _attrs$stateful = attrs.stateful,
        stateful = _attrs$stateful === undefined ? false : _attrs$stateful;


    _this.outputDim = outputDim;

    // keep this.activation and this.innerActivation for Bidirectional wrapper layer to use
    _this.activation = activation;
    _this.innerActivation = innerActivation;
    _this.activationFunc = activations[activation];
    _this.innerActivationFunc = activations[innerActivation];

    _this.returnSequences = returnSequences;
    _this.goBackwards = goBackwards;
    _this.stateful = stateful;

    // Layer weights specification
    _this.params = ['W_z', 'U_z', 'b_z', 'W_r', 'U_r', 'b_r', 'W_h', 'U_h', 'b_h'];
    return _this;
  }

  _createClass(GRU, [{
    key: 'call',


    /**
     * Method for layer computational logic
     * @param {Tensor} x
     * @returns {Tensor} x
     */
    value: function call(x) {
      var _this2 = this;

      var currentX = new _Tensor2.default([], [x.tensor.shape[1]]);

      var dimUpdateGate = this.weights['b_z'].tensor.shape[0];
      var dimResetGate = this.weights['b_r'].tensor.shape[0];
      var dimHiddenState = this.weights['b_h'].tensor.shape[0];

      var currentUpdateGateState = new _Tensor2.default([], [dimUpdateGate]);
      var tempXZ = new _Tensor2.default([], [dimUpdateGate]);
      var tempHZ = new _Tensor2.default([], [dimUpdateGate]);

      var currentResetGateState = new _Tensor2.default([], [dimResetGate]);
      var tempXR = new _Tensor2.default([], [dimResetGate]);
      var tempHR = new _Tensor2.default([], [dimResetGate]);

      var currentHiddenState = this.stateful && this.currentHiddenState ? this.currentHiddenState : new _Tensor2.default([], [dimHiddenState]);
      var tempXH = new _Tensor2.default([], [dimHiddenState]);
      var tempHH = new _Tensor2.default([], [dimHiddenState]);
      var previousHiddenState = new _Tensor2.default([], [dimHiddenState]);

      this.hiddenStateSequence = new _Tensor2.default([], [x.tensor.shape[0], dimHiddenState]);

      var _clearTemp = function _clearTemp() {
        var tempTensors = [tempXZ, tempHZ, tempXR, tempHR, tempXH, tempHH];
        tempTensors.forEach(function (temp) {
          return _ndarrayOps2.default.assigns(temp.tensor, 0);
        });
      };

      var _step = function _step() {
        _ndarrayOps2.default.assign(previousHiddenState.tensor, currentHiddenState.tensor);

        (0, _ndarrayBlasLevel.gemv)(1, _this2.weights['W_z'].tensor.transpose(1, 0), currentX.tensor, 1, tempXZ.tensor);
        (0, _ndarrayBlasLevel.gemv)(1, _this2.weights['U_z'].tensor.transpose(1, 0), previousHiddenState.tensor, 1, tempHZ.tensor);
        _this2._combine(currentUpdateGateState.tensor, tempXZ.tensor, tempHZ.tensor, _this2.weights['b_z'].tensor);
        _this2.innerActivationFunc(currentUpdateGateState);

        (0, _ndarrayBlasLevel.gemv)(1, _this2.weights['W_r'].tensor.transpose(1, 0), currentX.tensor, 1, tempXR.tensor);
        (0, _ndarrayBlasLevel.gemv)(1, _this2.weights['U_r'].tensor.transpose(1, 0), previousHiddenState.tensor, 1, tempHR.tensor);
        _this2._combine(currentResetGateState.tensor, tempXR.tensor, tempHR.tensor, _this2.weights['b_r'].tensor);
        _this2.innerActivationFunc(currentResetGateState);

        _ndarrayOps2.default.muleq(currentResetGateState.tensor, previousHiddenState.tensor);
        (0, _ndarrayBlasLevel.gemv)(1, _this2.weights['W_h'].tensor.transpose(1, 0), currentX.tensor, 1, tempXH.tensor);
        (0, _ndarrayBlasLevel.gemv)(1, _this2.weights['U_h'].tensor.transpose(1, 0), currentResetGateState.tensor, 1, tempHH.tensor);
        _this2._combine(currentHiddenState.tensor, tempXH.tensor, tempHH.tensor, _this2.weights['b_h'].tensor);
        _this2.activationFunc(currentHiddenState);

        _this2._update(currentHiddenState.tensor, previousHiddenState.tensor, currentUpdateGateState.tensor);
      };

      for (var i = 0, len = x.tensor.shape[0]; i < len; i++) {
        var inputIndex = this.goBackwards ? len - i - 1 : i;
        _ndarrayOps2.default.assign(currentX.tensor, x.tensor.pick(inputIndex, null));
        _clearTemp();
        _step();

        if (this.returnSequences) {
          _ndarrayOps2.default.assign(this.hiddenStateSequence.tensor.pick(i, null), currentHiddenState.tensor);
        }
      }

      if (this.returnSequences) {
        x.tensor = this.hiddenStateSequence.tensor;
      } else {
        x.tensor = currentHiddenState.tensor;
      }

      if (this.stateful) {
        this.currentHiddenState = currentHiddenState;
      }

      return x;
    }
  }]);

  return GRU;
}(_Layer3.default);

exports.default = GRU;
